### ISSUE TITLE
Very small regressor std causes huge regressor effect

### ISSUE THREAD

**Issue by JeremyKeustersML6** [facebook/prophet#1817]
I'm using the same set-up as explained in my earlier, non-related issue #1728 . This means that I'm first forecasting a group of items (eg. Group A) and then using the yearly seasonality output from this group in the individual item forecasts (eg. item A, item B, ..) as a regressor.

I'm however bumping into an issue that can happen in a specific condition.

I'm first doing a `model.fit()` with the following DataFrame:

ds | y | group_regressor
-- | -- | ----------------
2021-02-07 | 1 | 0.03551
2021-02-14 | 0 | 0.03381

(`group_regressor` is the `yearly` column from the group forecast)

After the fit, the calculated `std` and `mu` regressor values are as following (rounded): 
`std`: `0.001202`
`mu`: `0.034662`

I'm afterwards running a `model.predict()` using the generated future DataFrame:

ds | group_regressor
-- | -----------------
2021-02-21 | -0.03000
2021-02-28 | -0.09996
2021-03-07 | -0.17065
2021-03-14 | -0.26333
2021-03-21 | -0.32071
2021-03-28 | -0.18523
2021-04-04 | 0.25203
2021-04-11 | 0.87631
2021-04-18 | 1.37753
2021-04-25 | 1.50399
... | ...

However, the model uses the calculated `std` and `mu` values to standardize/scale the values (done at line 325 in `forecaster.py`), causing huge values for the regressor column:

ds | group_regressor
-- | -----------------
2021-02-21 | -53.79733
2021-02-28 | -122.00235
2021-03-07 | -170.81149
2021-03-14 | -247.91596
2021-03-21 | -295.65764
2021-03-28 | -182.94425
2021-04-04 | 180.84561
2021-04-11 | 700.21504
2021-04-18 | 1117.21481
2021-04-25 | 1222.42353
... | ...

The formula used on line 325 in `forecaster.py` is:
`scaled_value = (original_value - mu) / std`

Eg:
`scaled_value = (1.50399 - 0.034662) / 0.001202`
`scaled_value = 1222,40266` (which is ~ the value in the last row of the above table, not exact due to rounding)

So basically, because coincidentally the two regressor values (0.03551 & 0.03381) are very close to each other, the standard deviation is a very small number causing the scaled regressor values in the prediction phase to be huge (because of the division). On its turn, `yhat` values are also extremely large. Is this an expected behaviour? Or is something going wrong here? And what could be a possible workaround (except for inputting more data). I feel like the model should somehow take into account that this can happen and handle it? Thanks!

**Comment by JeremyKeustersML6** [comment]
Just a small update on this issue.

My first idea was to add more historical `group_regressor` values (as I have those) and use `n/a` for the corresponding `y` values. However, I quickly discovered that the rows with `n/a` values are dropped when fitting, so this doesn't help.

Something else I tried is setting the `standardize` argument in the `model.add_regressor()` method to `False`. This does seem to work, although I stumbled upon another issue, which happens regardless of the `standardize` argument.

When the `y` value goes down and the `regressor_group` value goes up during the `model.fit()` phase, the effect of the regressor seems to be 'flipped'/'inverted'. Let's take a monthly aggregated example with `standardize` on `False`:

I'm first doing a `model.fit()` with the following DataFrame:

ds | y | group_regressor
-- | -- | ----------------
2020-04-30 | 350 | -0.95155
2020-05-31 | 0 | -0.84554
2020-06-30 | 0 | -0.89737

(`group_regressor` is the `yearly` column from the group forecast)

I'm afterwards running a `model.predict()` using the generated future DataFrame (please note that I use a flat trend):

ds | group_regressor
-- | -----------------
2020-07-31 | -1.64005
2020-08-31 | -4.47732
2020-09-30 | 1.51688
2020-10-31 | 0.12901
2020-11-30 | 0.92416
2020-12-31 | -0.14061
2021-01-31 | -5.38624
2021-02-28 | -0.55279
2021-03-31 | 3.29045
2021-04-30 | -0.92985
2021-05-31 | -0.83532
2021-06-30 | -0.91118

The outcome of `model.predict()` is as follows:

ds          | trend    | group_regressor |  yhat
----------- | -------- | --------------- | ------
2020-07-31  | 48.18077 | 2.74764         | 180.56401
2020-08-31  | 48.18077 | 7.50100         | 409.58456
2020-09-30  | 48.18077 | -2.54127        | -74.25966
2020-10-31  | 48.18077 | -0.21613        | 37.76749
2020-11-30  | 48.18077 | -1.54828        | -26.41662
2020-12-31  | 48.18077 | 0.23557         | 59.53077
2021-01-31  | 48.18077 | 9.02374         | 482.95136
2021-02-28  | 48.18077 | 0.92611         | 92.80165
2021-03-31  | 48.18077 | -5.51259        | -217.42020
2021-04-30  | 48.18077 | 1.55782         | 123.23753
2021-05-31  | 48.18077 | 1.39945         | 115.60718
2021-06-30  | 48.18077 | 1.52653         | 121.73029

There are 2 important remarks about this forecast to be made:

**1. Inverse forecast**
The dates that should have the lowest prediction actually have the highest (inverse forecast). Take for example the date of `2021-01-31`: the `group_regressor` value before the prediction was `-5.38624`, which is the lowest value of all regressor values. After the `model.predict()`, we see that the value has become `9.02374`, which is suddenly the highest value from all. This causes on its turn the `yhat` to also be the highest of all values, while this is wrong of course.

As already said, I suppose this is happening because the numbers actually tell us that a higher regression value means a lower `y` value during the fitting of the model.

I'm noticing that the `model.beta` parameter in these cases is negative (eg. `-1.6753329`). Based on the example Stan code/formula that you listed in your paper, I assume that the negative `beta` also makes the `X` parameter negative and thus inverses the linear likelihood formula outcome.
Formula (p14 in the Prophet paper): `y ∼ normal((k + A * delta) .* t + (m + A * gamma) + X * beta, sigma);`

**2. Strange trend**
The trend is `48.18077`, which doesn't make sense as we're using a flat trend of 3 values (355, 0 and 0). One would assume that the trend of the flat trend would be `(355+0+0)/3=118.33`, as is the case when the `standardize` argument is on `auto`.

---

Do you have any suggestions on how to fix this issue, and could you also tell me if not standardizing these regressor values is a good/valid approach? Thanks in advance!

**Comment by bletham** [comment]
As you note, standardization can be disabled by passing in `standardize=False` when adding the regressor. While I suppose that could be helpful if the std is really small, practically, if the std is really small relative to the mean it means the regressor is basically constant and probably shouldn't be used anyway (won't be useful).

Prophet is going to give pretty pathological results on both of these time series because they are very short (especially relative to the forecast period) and don't really have any of the structure that Prophet tries to fit (piecewise trend, seasonality). With 3 datapoints, the model is going to be very badly overfit and basically what we're seeing here is that.

In the second dataset, y drops from 350 to 0 while the extra regressor increases from -0.95 to -0.87ish. So the model learns that an increase in extra regressor on the scale of +0.08 corresponds to a large decrease in y. Thus the negative value for beta. So in the future, the point with the most negative regressor is what we'll expect to have the most positive y. The most negative regressor in the `future` table is -5.38624 for 1-31. In the forecast dataframe produced by `predict`, what it shows is `beta*regressor` which is the actual contribution to y of the regressor. Because beta is negative, the large decrease in regressor value corresponds to a large increase in y, and we see yhat of 482.

I'd also generally expect to see the trend fall closer to the average of the data, but I think the regressor effect is just overfitting and the trend is just being put almost arbitrarily for the regressor. I'd try adding a whole lot of regularization to the regressor, like `add_regressor(name='group_regressor', prior_scale=0.01)`. (Default is 10). As you crank that down, you should see a lot less variability in the regressor effect, and the trend will probably migrate towards the mean.

A final note on the short histories - like I said Prophet will generally not do well on time series like these because they do not have the structure Prophet models (piecewise trend, seasonality). But even if it did I would be pretty hesitant about trying to forecast that far into the future from so little data. How far you can predict depends a lot on the stability of the time series, but I generally would avoid predicting out more than about 1/3rd of the length of the history unless I had some reason to believe that the time series was particularly stable.

**Comment by JeremyKeustersML6** [comment]
Hi Ben! Thanks for sharing your thoughts on this matter! Since there's no real 'fix' for this, I'm only using the group regressor once there's more data available. Although the forecast might still be a bit off, it at least fixes the 'exploding' forecast. I might experiment further in the future with the `prior_scale` as you say, to regularise it more.
